File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/FandCTyped.java
package spectable.common;

import java.util.ArrayList;
import java.util.List;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class FandCTyped {
    public int f;
    public int c;
    public String notes;

    public FandCTyped(int f, int c, String notes) {
        this.f = f;
        this.c = c;
        this.notes = notes;
    }

    public FandCTyped(FandCString s) {
        this.f = Integer.parseInt(s.f);
        this.c = Integer.parseInt(s.c);
        this.notes = s.notes;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof FandCTyped)) return false;
        FandCTyped that = (FandCTyped) o;
        return Objects.equals(f, that.f)
            && Objects.equals(c, that.c)
            && Objects.equals(notes, that.notes);
    }

    @Override
    public int hashCode() {
        return Objects.hash(f, c, notes);
    }

    public FandCString toFandCString() {
        return new FandCString(String.valueOf(f), String.valueOf(c), String.valueOf(notes));
    }

    public static List<FandCTyped> fromStringList(List<FandCString> list) {
        List<FandCTyped> result = new ArrayList<>();
        for (FandCString s : list) result.add(new FandCTyped(s));
        return result;
    }

    public static List<FandCString> toStringList(List<FandCTyped> list) {
        List<FandCString> result = new ArrayList<>();
        for (FandCTyped t : list) result.add(t.toFandCString());
        return result;
    }

    /** Internal plumbing for {@link Json}; use toJSON() for JSON text. */
    public Map<String, Object> toJsonValue() {
        Map<String, Object> m = new LinkedHashMap<String, Object>();
        m.put("f", f);
        m.put("c", c);
        m.put("notes", notes);
        return m;
    }

    public String toJSON() { return Json.write(toJsonValue()); }

    /** Internal plumbing for {@link Json}; use fromJSON(String) for JSON text. */
    public static FandCTyped fromJsonValue(Map<String, Object> m) {
        if (m == null) return null;
        return new FandCTyped(
            Json.asInt(Json.require(m, "f"), "f"),
            Json.asInt(Json.require(m, "c"), "c"),
            Json.asString(Json.require(m, "notes"), "notes"));
    }

    public static FandCTyped fromJSON(String json) {
        return fromJsonValue(Json.parseObject(json));
    }

    public static String toJSONList(List<FandCTyped> list) {
        List<Object> arr = new ArrayList<Object>();
        if (list != null)
            for (FandCTyped item : list) arr.add(item == null ? null : item.toJsonValue());
        return Json.write(arr);
    }

    public static List<FandCTyped> fromJSONList(String json) {
        List<FandCTyped> result = new ArrayList<FandCTyped>();
        for (Object e : Json.parseArray(json))
            result.add(fromJsonValue(Json.asObject(e, "FandCTyped")));
        return result;
    }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/FandCTyped.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using System;
using System.Buffers;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.Json;
using production;

    public class FandCTyped
    {
        public int f;
        public int c;
        public string notes;

        public FandCTyped(int f, int c, string notes)
        {
            this.f = f;
            this.c = c;
            this.notes = notes;
        }

        public FandCString ToFandCString()
        {
            return new FandCString(
                Json.ToText(this.f),
                Json.ToText(this.c),
                Json.ToText(this.notes)
            );
        }

        public static List<FandCString> ToStringList(List<FandCTyped> list)
        {
            var result = new List<FandCString>();
            foreach (var t in list) result.Add(t.ToFandCString());
            return result;
        }

        public static List<FandCTyped> FromStringList(List<FandCString> list)
        {
            var result = new List<FandCTyped>();
            foreach (var s in list) result.Add(s.ToFandCTyped());
            return result;
        }

        public void WriteJson(Utf8JsonWriter w)
        {
            w.WriteStartObject();
            w.WriteNumber("f", this.f);
            w.WriteNumber("c", this.c);
            w.WriteString("notes", this.notes);
            w.WriteEndObject();
        }

        public string ToJSON()
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer)) { WriteJson(w); }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static FandCTyped FromJsonElement(JsonElement m)
        {
            return new FandCTyped(
                Json.AsInt(Json.Require(m, "f"), "f"),
                Json.AsInt(Json.Require(m, "c"), "c"),
                Json.AsString(Json.Require(m, "notes"), "notes")
            );
        }

        public static FandCTyped FromJSON(string json)
        {
            return FromJsonElement(Json.Parse(json));
        }

        public static string ToJSONList(List<FandCTyped> list)
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer))
            {
                w.WriteStartArray();
                foreach (var item in list) item.WriteJson(w);
                w.WriteEndArray();
            }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static List<FandCTyped> FromJSONList(string json)
        {
            var result = new List<FandCTyped>();
            var root = Json.Parse(json);
            Json.RequireArray(root, "FandCTyped");
            foreach (var e in root.EnumerateArray()) result.Add(FromJsonElement(e));
            return result;
        }

        public override string ToString()
        {
            return $"F={f}, C={c}, Notes={notes}";
        }

        public override bool Equals(object? obj)
        {
            if (obj is not FandCTyped other) return false;
            return object.Equals(this.f, other.f)
                && object.Equals(this.c, other.c)
                && object.Equals(this.notes, other.notes);
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.f);
            h.Add(this.c);
            h.Add(this.notes);
            return h.ToHashCode();
        }
    }
}
